src/app/loadout-builder/auto-optimizer.ts

```diff
@@ -0,0 +1,434 @@
+  | { tag: 'finished'; result: AutoOptimizationResult }
+  | { tag: 'error'; error: LoadoutError };
+
+/**
+ * We ran the auto-opt process for a set and this was the result.
+ */
+export const enum AutoOptimizationResult {
+  /** This build is the best it could be. Hooray! */
+  Nothing = 0,
+  /** There's a set that is better or equal in all stats and better in at least one stat. */
+  StrongBetterSet,
+  /**
+   * There's a set that matches the original constraints and compares more favorably
+   * (higher tier total, or equal tier total and better in highest-priority stat, or...)
+   */
+  WeakBetterSet,
+}
+
+/**
+ * A loadout was ineligible for auto-optimizing.
+ */
+export const enum LoadoutError {
+  /** The armor set did not have 5 equipped armor items, so we can't come up with stats to compare against. */
+  NotAFullArmorSet = 1,
+  /** The armor set specifies an exotic but the loadout doesn't even have that exotic, so that's cheating! */
+  DoesNotRespectExotic,
+  /** The armor set does not fit all requested mods in the first place, so there's no meaningful comparison with other sets. */
+  ModsDontFit,
+  /** The loadout's search query is invalid */
+  BadSearchQuery,
+}
+
+function matchesExoticArmorHash(exoticArmorHash: number | undefined, exotic: DimItem | undefined) {
+  if (exoticArmorHash === LOCKED_EXOTIC_NO_EXOTIC) {
+    return !exotic;
+  } else if (exoticArmorHash === LOCKED_EXOTIC_ANY_EXOTIC) {
+    return Boolean(exotic);
+  } else {
+    return exoticArmorHash === undefined || exoticArmorHash === exotic?.hash;
+  }
+}
+
+export interface AutoOptimizationReport {
+  [loadoutId: string]: ArmorSetResult;
+}
+
+function extractOptimizationParameters(
+  autoOptContext: AutoOptimizationContext,
+  loadout: Loadout
+): AutoOptimizationParameters | LoadoutError {
+  const loadoutParameters: LoadoutParameters = {
+    ...autoOptContext.savedLoLoadoutParameters,
+    ...loadout.parameters,
+  };
+
+  let searchFilter;
+  if (loadoutParameters.query) {
+    const { valid } = autoOptContext.validateQuery(loadoutParameters.query);
+    if (!valid) {
+      return LoadoutError.BadSearchQuery;
+    }
+    searchFilter = autoOptContext.filterFactory(loadoutParameters.query);
+  } else {
+    searchFilter = _.stubTrue;
+  }
+
+  const [items, subclass] = getItemsAndSubclassFromLoadout(
+    autoOptContext.itemCreationContext,
+    loadout.items,
+    autoOptContext.store,
+    autoOptContext.allItems
+  );
+
+  const armorEnergyRules: ArmorEnergyRules = {
+    minItemEnergy: MIN_LO_ITEM_ENERGY,
+    assumeArmorMasterwork: loadoutParameters.assumeArmorMasterwork!,
+  };
+
+  const armorItems = items
+    .filter((i) => i.loadoutItem.equip && i.item.bucket.inArmor && i.item.energy)
+    .map((i) => i.item);
+  // A loadout must have 5 equipped armor 2.0 items for comparisons
+  // between LO runs and the current loadout to be meaningful
+  if (armorItems.length !== 5) {
+    return LoadoutError.NotAFullArmorSet;
+  }
+
```

Comment:
Maybe it should validate that the armor items match the search query too?

Reply:
Kind of tricky with the special behavior LO has for ignoring a search query in some cases...